File: solutions/cpp/0008.cpp
class Solution {
 public:
  int myAtoi(string str) {
    long ans = 0;
    bool isNegative = false;
    int j = 0;

    while (j < str.length() && str[j] == ' ') ++j;
    if (j == str.length()) return 0;
    if (str[j] == '-') {
      isNegative = true;
      ++j;
    } else if (str[j] == '+') {
      ++j;
    }

    for (int i = j; i < str.length(); ++i) {
      if (str[i] < '0' || str[i] > '9')
        break;
      else {
        ans = ans * 10 + (str[i] - '0');
        if (isNegative && -ans <= INT_MIN) return INT_MIN;
        if (!isNegative && ans >= INT_MAX) return INT_MAX;
      }
    }

    return isNegative ? -ans : ans;
  }
};

File: solutions/java/0008.java
class Solution {
  public int myAtoi(String str) {
    long ans = 0;
    str = str.strip();

    if (str.isEmpty())
      return 0;

    boolean isNegative = str.charAt(0) == '-';
    if (str.charAt(0) == '-' || str.charAt(0) == '+')
      str = str.substring(1, str.length());

    for (char c : str.toCharArray()) {
      if (!Character.isDigit(c))
        break;
      ans = ans * 10 + (c - '0');
      if (isNegative && -ans <= Integer.MIN_VALUE)
        return Integer.MIN_VALUE;
      if (!isNegative && ans >= Integer.MAX_VALUE)
        return Integer.MAX_VALUE;
    }

    return isNegative ? (int) -ans : (int) ans;
  }
}